experimental/metal/pipeline_layout.m

```diff
@@ -0,0 +1,180 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "experimental/metal/pipeline_layout.h"
+
+#include <stddef.h>
+
+#include "iree/base/api.h"
+#include "iree/base/tracing.h"
+
+//===------------------------------------------------------------------------------------------===//
+// iree_hal_metal_descriptor_set_layout_t
+//===------------------------------------------------------------------------------------------===//
+
+typedef struct iree_hal_metal_descriptor_set_layout_t {
+  // Abstract resource used for injecting reference counting and vtable; must be at offset 0.
+  iree_hal_resource_t resource;
+
+  iree_allocator_t host_allocator;
+
+  iree_host_size_t binding_count;
+  iree_hal_descriptor_set_layout_binding_t bindings[];
+} iree_hal_metal_descriptor_set_layout_t;
+
+static const iree_hal_descriptor_set_layout_vtable_t iree_hal_metal_descriptor_set_layout_vtable;
+
+static iree_hal_metal_descriptor_set_layout_t* iree_hal_metal_descriptor_set_layout_cast(
+    iree_hal_descriptor_set_layout_t* base_value) {
+  IREE_HAL_ASSERT_TYPE(base_value, &iree_hal_metal_descriptor_set_layout_vtable);
+  return (iree_hal_metal_descriptor_set_layout_t*)base_value;
+}
+
+iree_status_t iree_hal_metal_descriptor_set_layout_create(
+    iree_allocator_t host_allocator, iree_hal_descriptor_set_layout_flags_t flags,
```

Comment:
by convention host_allocator immediately precedes the thing its used for - in creation functions that means `(....., host_allocator, out_resource)` so it should be moved before out_descriptor_set_layout